gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java

```diff
@@ -95,6 +97,56 @@ public static Builder newBuilder() {
         .setUsingGDCH(false);
   }
 
+  /** Check that the User configured universe domain matches the Credentials' universe domain */
+  public boolean hasValidUniverseDomain(Credentials credentials) throws IOException {
+    if (usingGDCH()) {
+      // GDC-H has no universe domain, always return true.
+      return true;
+    } else if (credentials == null) {
+      // If passed with NoCredentialsProvider, match with GDU
+      return resolvedUniverseDomain().equals(Credentials.GOOGLE_DEFAULT_UNIVERSE);
+    }
+    return resolvedUniverseDomain().equals(credentials.getUniverseDomain());
+  }
+
+  public EndpointContext merge(EndpointContext input) {
+    if (input == null) {
+      return this;
+    }
+    Builder builder = this.toBuilder();
+    String serviceName = input.serviceName();
+    if (serviceName != null) {
+      builder.setServiceName(serviceName);
+    }
+    String universeDomain = input.universeDomain();
+    if (universeDomain != null) {
+      builder.setUniverseDomain(universeDomain);
+    }
+    String clientSettingsEndpoint = input.clientSettingsEndpoint();
+    if (clientSettingsEndpoint != null) {
+      builder.setClientSettingsEndpoint(clientSettingsEndpoint);
+    }
+    String transportChannelProviderEndpoint = input.transportChannelProviderEndpoint();
+    if (transportChannelProviderEndpoint != null) {
+      builder.setTransportChannelProviderEndpoint(transportChannelProviderEndpoint);
+    }
+    String mtlsEndpoint = input.mtlsEndpoint();
+    if (mtlsEndpoint != null) {
+      builder.setMtlsEndpoint(mtlsEndpoint);
+    }
+    builder.setSwitchToMtlsEndpointAllowed(input.switchToMtlsEndpointAllowed());
+    MtlsProvider mtlsProvider = input.mtlsProvider();
+    if (mtlsEndpoint != null) {
+      builder.setMtlsProvider(mtlsProvider);
+    }
+    builder.setUsingGDCH(input.usingGDCH());
+    try {
+      return builder.build();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
```

Comment:
Throw a RuntimeException so the checked IOException does not get bubbled up to calling methods.